### DISCUSSION TITLE
How to refactor a method based on some other parts of the file

### DISCUSSION THREAD

**Discussion by vaceletm** [rectorphp/rector#7822] [Q&A] [answered]
Hi there,

First of all, thanks for the awesome work done on Rector.

I've been looking for examples of refactoring that needs some kind of memory/persistence but I failed to find something relevant (probably because I don't really know how to state my problem...).

Here it is, I've got some classes that looks like:
```php
class Stuff {
    ...

    public function getHooksAndCallbacks() {
        $this->addHook('some_event1', 'someString1');
        $this->addHook('some_event2', 'someString2');
    }

    ...

    public function someString1(...) {}

    public function someString2(...) {}

}
```

I want to inject Attributes on `someString1()` & `someString2()` method based on the content of `getHooksAndCallbacks()`.

Ideal result would be
```php
class Stuff {
    ...

    #[SomeAttribute('some_event1')]
    public function someString1(...) {}

    #[SomeAttribute('some_event2')]
    public function someString2(...) {}

}
```

I would need some kind of double processing of the file I guess but there is maybe a better approach. Any suggestion ?

**Comment by TomasVotruba** [comment, marked as answer]
Hi, you can process it just once. Hook into the `Class_` node and handle it there.

**Reply by vaceletm** [reply]
Ok, I fiddled a bit with the proposal and it fits perfectly. Thanks for the proposal.

I'll polish a bit the code and link it here for people that are looking for examples.

**Reply by vaceletm** [reply]
I managed to do my transformation, the whole code is here for those looking for examples: https://gerrit.tuleap.net/plugins/gitiles/tuleap/+/refs/changes/20/28020/1/tools/utils/Rector/ConvertAddHookToAttributesRule.php
